plugins/offline-storage.js

```diff
@@ -0,0 +1,61 @@
+/**
+ * Offline Storage plugin
+ * 
+ * Stores errors failed to get send and try to send them when
+ * Networkf comes back or on init
+ */
+
+var offlineStorageKey = 'raven-js-offline-queue';
+
+function offlineStoragePlugin(Raven, options) {
+  options = options || {};
+
+  function processOfflineQueue() {
+    // Let's stop here if there's no connection
+    if (!navigator.onLine) {
+      return;
+    }
+
+    try {
+      // Get the queue
+      var queue = JSON.parse(localStorage.getItem(offlineStorageKey)) || [];
+
+      // Store an empty queue. If processing these one fails they get back to the queue
+      localStorage.removeItem(offlineStorageKey);
```

Comment:
multiple localStorage operations are not atomic. It is possible for multiple tabs to receive this queue before the `removeItem`. This needs to be inside a lock. See more info here:
https://medium.engineering/wait-dont-touch-that-a211832adc3a
http://balpha.de/2012/03/javascript-concurrency-and-locking-the-html5-localstorage/

Reply:
https://www.npmjs.com/package/localstorage-fifo ?

Reply:
@graingert I don't believe that library handles atomic operations between multiple browser windows. It just removes oldest item when localstorage when it's full

Reply:
Here is a link to a library I created to solve this problem. There are probably other libraries out there as well. Also feel free to just copy the library code into this repo if you don't want another dependency.

https://github.com/taylorhakes/localstorage-lock/

Reply:
Storing an entire array of the logs at once can be quite a poor usage of memory. On hybrid apps there is a very limit memory limit and if there are a lot of logs, that can build up. Possibly using indexdb or an option for a custom storage adapter would allow for more memory efficient ways of storing.